File: LeetCodeProblemsCMake/Problems/LeetCode3304FindTheKthCharacterInStringGameI.cpp
//
// Created by saeed on 7/3/25.
//

#include "LeetCode3304FindTheKthCharacterInStringGameI.h"

#include <iostream>
#include <ostream>
#include <string>
using  namespace std;
void GetString(string current,int indexBefore,int k,char & result) {
    int length=current.length();
    if (length+indexBefore>=k) {
        k-=indexBefore;
        char target=current[k-1];
        result=target<'z'?target+1:'a';
        return;
    }
    for (int i=0;i<length;i++) {
        char target=current[i];
        target=target<'z'?target+1:'a';
        current+=target;
    }
    indexBefore+=length;
    GetString(current,indexBefore,k,result);
}
char LeetCode3304FindTheKthCharacterInStringGameI::kthCharacter(int k) {
    if (k==1)
        return 'a';
    char result=0;
    GetString("a",1,k,result);
    return result;
}

void LeetCode3304FindTheKthCharacterInStringGameI::Run() {
    int k=1;
    cout<< kthCharacter(k)<<endl;
}

File: LeetCodeProblemsGo/LeetCode3304FindTheKthCharacterInStringGameI/LeetCode3304FindTheKthCharacterInStringGameI.go
package leetcode3304findthekthcharacterinstringgamei

func kthCharacter(k int) byte {
	if k == 1 {
		return 'a'
	}
	var result byte
	getString("a", 1, k, &result)
	return result
}

func getString(current string, indexBefore, k int, result *byte) {
	length := len(current)
	if length+indexBefore >= k {
		k -= indexBefore
		target := current[k-1]
		target++
		if target > 'z' {
			target = 'a'
		}
		*result = target
		return
	}
	for i := 0; i < length; i++ {
		target := current[i]
		target++
		if target > 'z' {
			target = 'a'
		}
		current += string(rune(target))
	}
	indexBefore += length
	getString(current, indexBefore, k, result)
}
